Coral/Coral.js

```diff
@@ -0,0 +1,312 @@
+    this.position = { x, y };
+    this.dimensions = { x: width, y: height };
+  }
+
+  clamp(point) {
+    const { x, y } = point;
+
+    return {
+      x: clamp(x, this.position.x, this.position.x + this.dimensions.x),
+      y: clamp(y, this.position.y, this.position.y + this.dimensions.y),
+    };
+  }
+}
+
+class ControlPoint {
+  constructor(x, y, dx, dy) {
+    this._position = { x, y };
+    this._tangent = { x: dx, y: dy };
+
+    this.forward_point = add(this._position, this._tangent);
+    this.backward_point = sub(this._position, this._tangent);
+  }
+
+  get position() {
+    return this._position;
+  }
+
+  set position(value) {
+    this._position = value;
+    this.forward_point = add(this._position, this._tangent);
+    this.backward_point = sub(this._position, this._tangent);
+  }
+
+  get tangent() {
+    return this._tangent;
+  }
+
+  set tangent(value) {
+    this._tangent = value;
+    this.forward_point = add(this._position, this._tangent);
+    this.backward_point = sub(this._position, this._tangent);
+  }
+}
+
+class Spline {
+  constructor(control_points) {
+    this.control_points = control_points;
+  }
+
+  log_params() {
+    console.log("current control points:");
+    for (const point of this.control_points) {
+      console.log(
+        "p:",
+        point.position.x.toPrecision(3),
+        point.position.y.toPrecision(3)
+      );
+      console.log(
+        "v:",
+        point.tangent.x.toPrecision(3),
+        point.tangent.y.toPrecision(3)
+      );
+    }
+  }
+}
+
+const SELECT_RADIUS = 8;
+const SELECT_RADIUS_SQR = SELECT_RADIUS * SELECT_RADIUS;
+
+class InteractiveVertex {
+  constructor(control_point, constraint) {
+    this.control_point = control_point;
+    this.constraint = constraint;
+  }
+
+  get position() {
+    return this.control_point.position;
+  }
+
+  is_hovering(mouse) {
+    const position_world = uv_to_world(this.control_point.position);
+    const dist_sqr = norm(sub(mouse, position_world));
+
+    return dist_sqr < SELECT_RADIUS_SQR;
+  }
+
+  move(uv) {
+    this.control_point.position = this.constraint.clamp(uv);
+  }
+}
+
+class InteractiveTangent {
+  constructor(control_point, constraint) {
+    this.control_point = control_point;
+    this.constraint = constraint;
+  }
+
+  get tip() {
+    return this.control_point.forward_point;
+  }
+
+  is_hovering(mouse) {
+    const position_world = uv_to_world(this.control_point.forward_point);
+    const dist_sqr = norm(sub(mouse, position_world));
+
+    return dist_sqr < SELECT_RADIUS_SQR;
+  }
+
+  move(uv) {
+    const tangent = sub(uv, this.control_point.position);
+    this.control_point.tangent = this.constraint.clamp(tangent);
+  }
+}
+
+const SPLINE = new Spline([
+  new ControlPoint(0.75, 0.0, -0.21, 0.219),
+  new ControlPoint(0.75, 0.5, 0.09, 0.204),
+  new ControlPoint(0.5, 0.934, -0.185, 0.055),
+  new ControlPoint(0.25, 0.5, 0.11, -0.291),
+  new ControlPoint(0.25, 0.0, -0.095, -0.176),
```

Comment:
This 5th control point isn't actually part of the tile (it's part of the neighboring tile), but it's needed if the curve is not closed.